accepted/future-releases/extension-types/feature-specification.md

```diff
@@ -0,0 +1,1212 @@
+representation are bound to `o`, and the type variables of `V` are
+bound to the actual values of
+<code>T<sub>1</sub>, .. T<sub>s</sub></code>.
+If the body completes returning an object `o2` then the invocation
+evaluates to `o2`. If the body throws an object and a stack trace
+then the invocation completes throwing the same object and stack
+trace.
+
+Otherwise, if `args` is omitted and _Dm_ is a method, the invocation
+evaluates to a closurization of _Dm_ where
+`this` and the name of the representation are bound to `o`, and the
+type variables of `V` are bound to the actual values of
+<code>T<sub>1</sub>, .. T<sub>s</sub></code>.
+The operator `==` of the closurization returns true if and only if the
+operand is the same object.
+
+*Loosely said, these function objects simply use the equality inherited
+from `Object`, there are no special exceptions. Note that we can tear off
+the same method from the same extension type with the same representation
+object twice, and still get different behavior, because the extension type
+had different actual type arguments. Hence, we can not consider two
+extension type method tear-offs equal just because they have the same
+receiver.*
+
+Otherwise, the following is known: `args` is included, and _Dm_ is a
+method. The invocation proceeds to evaluate `args` to an actual
+argument list `args1`. Then it executes the body of _Dm_ in an
+environment where `this` and the name of the representation are bound
+to `o`, the type variables of `V` are bound to the actual values of
+<code>T<sub>1</sub>, .. T<sub>s</sub></code>,
+and the formal parameters of `m` are bound to `args1` in the same way
+that they would be bound for a normal function call.
+If the body completes returning an object `o2` then the invocation
+evaluates to `o2`. If the body throws an object and a stack trace
+then the invocation completes throwing the same object and stack
+trace.
+
+
+## Static Analysis of Extension Types
+
+Assume that
+<code>T<sub>1</sub>, .. T<sub>s</sub></code>
+are types, and `V` resolves to an extension type declaration of the
+following form:
+
+```dart
+extension type V<X1 extends B1, .. Xs extends Bs>(T id) ... {
+  ... // Members.
+}
+```
+
+It is then allowed to use
+<code>V&lt;T<sub>1</sub>, .. T<sub>s</sub>&gt;</code>
+as a type.
+
+*For example, it can occur as the declared type of a variable or parameter,
+as the return type of a function or getter, as a type argument in a type,
+as the representation type of an extension type declaration, as the on-type
+of an extension declaration, as the type in the `onPart` of a try/catch
+statement, in a type test `o is V<...>`, in a type cast `o as V<...>`, or
+as the body of a type alias. It is also allowed to create a new instance
+where one or more extension types occur as type arguments (e.g.,
+`List<V>.empty()`).*
+
+A compile-time error occurs if the type
+<code>V&lt;T<sub>1</sub>, .. T<sub>s</sub>&gt;</code>
+is not regular-bounded.
```

Comment:
So we say that it's "allowed" when it's well-formed (correct number of type arguments).
Presumably also as a raw type.

Then later we check if it's regular-bounded, and complain if it's not.

Why not say "compile-time error" first, instead of "allowed":

> Given a term of the form <Code>V\<T<sub>1</sub>, .. T<sub>s</sub>\></code>
> where `V` resolves to an extension type declaration of the form:
> ```dart
> extension type V<X1 extends B1, .. Xs extends Br>(T id) ... {
>   ... // Members.
> }
> ```
> it's a compile-time error if <Code>s &ne; r</code>.

Then we don't get into saying "allowed", without saying what happens if it's *not* allowed. Everything is allowed, if it's not a compile-time error.

Reply:
I don't know where it says "allowed" with that meaning. The only "allowed" I could spot now has to do with the ability to use `C<T1 .. Tk>` as a type, and that doesn't imply anything about that being a correct type. 'Regular-bounded' implies that the number of type arguments is the same as the number of type parameters, and the list of type arguments satisfies the list of bounds (it doesn't make sense to say that each actual type argument satisfies its bound, because some type parameters can be used in some bounds).

Reply:
Ah, reading this one more time, I could see the problem. Revised.